backfiller/backfiller/main.py

```diff
@@ -0,0 +1,227 @@
+
+		resp.raise_for_status()
+
+		with open(temp_path, 'w') as f:
+			for chunk in resp.iter_content(8192):
+				f.write(chunk)
+
+	#try to get rid of the temp file if an exception is raised.
+	except Exception:
+		if os.path.exists(temp_path):
+			os.remove(temp_path)
+		raise
+
+	common.rename(temp_path, path)
+
+
+def backfill(base_dir, stream, variants, hours=None, nodes=None, start=None,
+		 stop=None, order=None):
+	"""Loop over nodes backfilling from each.
+
+	Backfill from node/stream/variants to base_dir/stream/variants for each node in nodes. If nodes is None, use get_nodes() to get a list of nodes to backfill from. Passes hours, start, stop and order to backfill_node to control which hours are backfilled and in which order. By default all hours are backfilled. If backfilling from a node raises an exception, this just goes onto the next node."""
+
+	if nodes is None:
+		nodes = get_nodes()
+
+	#ideally do this in parallel
+	for node in nodes:
+		try:
+			backfill_node(base_dir, node, stream, variants, hours, start, stop, order=order)
+		except Exception as e:
+			print node, e
+
+
+def is_iterable(x):
+	"""Test whether input is iterable."""
+	try:
+ 		iter(x)
+	except TypeError:
+		return False
+	return True
+
+
+def backfill_node(base_dir, node, stream, variants, hours=None, start=None,
+		stop=None, order=None, recent_cutoff=60):
+	"""Backfill from remote node.
+
+	Backfill from node/stream/variants to base_dir/stream/variants.
+
+	Keyword arguments:
+	hours -- If None (default), backfill all available hours. If iterable, backfill only hours in iterable. Otherwise backfill the last N hours, starting with the lastest.
+	start -- Only backfill hours starting after or equal to this datetime object. If None (default), backfill all hours.
+	stop -- Only backfill hours starting before or equal to this datetime object. If None (default), backfill all hours.
+	order -- If 'random', randomise the order of hours. If 'forward', sort the hours in acceding order. If 'reverse', sort the hours in descending order. Otherwise, do not change the order of hours (default).
+	recent_cutoff -- Skip backfilling segments younger than this number of seconds to prioritise letting the downloader grab these segments."""
+
+	if hours is None:
+		hours = list_remote_hours(node, stream, variant)
+	elif is_iterable(hours):
+		None
+	else:
+		n_hours = hours
+		if n_hours < 1:
+			raise ValueError('Number of hours has to be 1 or greater')
+		now = datetime.datetime.utcnow()
+		hours = [(now - i * timedelta(hours=1)).strftime(HOUR_FMT) for i in range(n_hours)]
+
+	if start is not None:
+		hours = [hour for hour in hours if hour >= start]
+	if stop is not None:
+		hours = [hour for hour in hours if hour <= stop]
+
+	# useful if running in parallel and expecting to fetch a segments from
+	# multiple hours (say on start up) so that you don't try to backfill the
+	# same hour at the same time 
+	if order == 'random':
+		hours = random.shuffle(hours)
```

Comment:
bug here. `random.shuffle` returns None, and shuffles the list in-place. stupid, i know.